File: source/spring-petclinic-rest/src/main/java/com/spring_petclinic/spring_petclinic_rest/application/rest/OwnerRestController.java
package com.spring_petclinic.spring_petclinic_rest.application.rest;

import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import com.spring_petclinic.spring_petclinic_rest.application.services.OwnerRestService;
import com.spring_petclinic.spring_petclinic_rest.application.models.OwnerDTO;
import com.spring_petclinic.spring_petclinic_rest.application.models.OwnerCreateDTO;
import com.spring_petclinic.spring_petclinic_rest.application.models.OwnerUpdateDTO;
import java.util.List;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import io.swagger.v3.oas.annotations.Parameter;

@RestController
@RequestMapping("/api/owners")
@Tag(name = "OwnerRestController")
@AllArgsConstructor
public class OwnerRestController {
    private final OwnerRestService ownerRestService;

    @GetMapping
    @Operation(summary = "getOwners")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "Returns the specified List<OwnerDTO>.") })
    public ResponseEntity<List<OwnerDTO>> getOwners() {
        final List<OwnerDTO> result = ownerRestService.getOwners();
        if (result.isEmpty()) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
        return new ResponseEntity<>(result, HttpStatus.OK);
    }

    @ResponseStatus(HttpStatus.OK)
    @PostMapping
    @Operation(summary = "addOwner")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "201", description = "Successfully created."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred.") })
    public void addOwner(@Parameter(required = true) @RequestBody OwnerCreateDTO dto) {
        ownerRestService.addOwner(dto);
    }

    @GetMapping(path = "/{ownerId}")
    @Operation(summary = "getOwner")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "Returns the specified OwnerDTO."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred."),
        @ApiResponse(responseCode = "404", description = "Can\'t find an OwnerDTO with the parameters provided.") })
    public ResponseEntity<OwnerDTO> getOwner(@Parameter(required = true) @PathVariable(value = "ownerId") int ownerId) {
        final OwnerDTO result = ownerRestService.getOwner(ownerId);
        if (result == null) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
        return new ResponseEntity<>(result, HttpStatus.OK);
    }

    @ResponseStatus(HttpStatus.OK)
    @PutMapping(path = "/{ownerId}")
    @Operation(summary = "updateOwner")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "204", description = "Successfully updated."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred.") })
    public void updateOwner(@Parameter(required = true) @PathVariable(value = "ownerId") int ownerId, @Parameter(required = true) @RequestBody OwnerUpdateDTO dto) {
        ownerRestService.updateOwner(ownerId, dto);
    }

    @ResponseStatus(HttpStatus.OK)
    @DeleteMapping(path = "/{ownerId}")
    @Operation(summary = "deleteOwner")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "Successfully deleted."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred.") })
    public void deleteOwner(@Parameter(required = true) @PathVariable(value = "ownerId") int ownerId) {
        ownerRestService.deleteOwner(ownerId);
    }

    @GetMapping(path = "/*/lastname/{lastName}")
    @Operation(summary = "getOwnersList")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "Returns the specified List<OwnerDTO>."),
        @ApiResponse(responseCode = "400", description = "One or more validation errors have occurred.") })
    public ResponseEntity<List<OwnerDTO>> getOwnersList(@Parameter(required = true) @PathVariable(value = "lastName") String lastName) {
        final List<OwnerDTO> result = ownerRestService.getOwnersList(lastName);
        if (result.isEmpty()) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
        return new ResponseEntity<>(result, HttpStatus.OK);
    }
}

File: source/dotnet-petclinic-rest/PetClinic.Api/Controllers/OwnerRestController.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Transactions;
using Intent.RoslynWeaver.Attributes;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using PetClinic.Application;
using PetClinic.Application.Dtos;
using PetClinic.Application.Interfaces;
using PetClinic.Domain.Common.Interfaces;
using PetClinic.Infrastructure.Persistence;

[assembly: DefaultIntentManaged(Mode.Fully)]
[assembly: IntentTemplate("Intent.AspNetCore.Controllers.Controller", Version = "1.0")]

namespace PetClinic.Api.Controllers
{
    [ApiController]
    [Route("api/owners")]
    public class OwnerRestController : ControllerBase
    {
        private readonly IOwnerService _appService;
        private readonly IUnitOfWork _unitOfWork;

        public OwnerRestController(IOwnerService appService, IUnitOfWork unitOfWork)
        {
            _appService = appService ?? throw new ArgumentNullException(nameof(appService));
            _unitOfWork = unitOfWork ?? throw new ArgumentNullException(nameof(unitOfWork));
        }

        /// <summary>
        /// </summary>
        /// <response code="200">Returns the specified List&lt;OwnerDTO&gt;.</response>
        [HttpGet]
        [ProducesResponseType(typeof(List<OwnerDTO>), StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
        public async Task<ActionResult<List<OwnerDTO>>> GetOwners(CancellationToken cancellationToken)
        {
            var result = default(List<OwnerDTO>);
            result = await _appService.GetOwners();
            return Ok(result);
        }

        /// <summary>
        /// </summary>
        /// <response code="201">Successfully created.</response>
        /// <response code="400">One or more validation errors have occurred.</response>
        [HttpPost]
        [ProducesResponseType(StatusCodes.Status201Created)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
        public async Task<ActionResult> AddOwner([FromBody] OwnerCreateDTO dto, CancellationToken cancellationToken)
        {
            using (var transaction = new TransactionScope(TransactionScopeOption.Required,
                new TransactionOptions() { IsolationLevel = IsolationLevel.ReadCommitted }, TransactionScopeAsyncFlowOption.Enabled))
            {
                await _appService.AddOwner(dto);
                await _unitOfWork.SaveChangesAsync(cancellationToken);
                transaction.Complete();
            }
            return Created(string.Empty, null);
        }

        /// <summary>
        /// </summary>
        /// <response code="200">Returns the specified OwnerDTO.</response>
        /// <response code="400">One or more validation errors have occurred.</response>
        /// <response code="404">Can't find an OwnerDTO with the parameters provided.</response>
        [HttpGet("{ownerId}")]
        [ProducesResponseType(typeof(OwnerDTO), StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
        public async Task<ActionResult<OwnerDTO>> GetOwner([FromRoute] int ownerId, CancellationToken cancellationToken)
        {
            var result = default(OwnerDTO);
            result = await _appService.GetOwner(ownerId);
            return result != null ? Ok(result) : NotFound();
        }

        /// <summary>
        /// </summary>
        /// <response code="204">Successfully updated.</response>
        /// <response code="400">One or more validation errors have occurred.</response>
        [HttpPut("{ownerId}")]
        [ProducesResponseType(StatusCodes.Status204NoContent)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
        public async Task<ActionResult> UpdateOwner(
            [FromRoute] int ownerId,
            [FromBody] OwnerUpdateDTO dto,
            CancellationToken cancellationToken)
        {
            using (var transaction = new TransactionScope(TransactionScopeOption.Required,
                new TransactionOptions() { IsolationLevel = IsolationLevel.ReadCommitted }, TransactionScopeAsyncFlowOption.Enabled))
            {
                await _appService.UpdateOwner(ownerId, dto);
                await _unitOfWork.SaveChangesAsync(cancellationToken);
                transaction.Complete();
            }
            return NoContent();
        }

        /// <summary>
        /// </summary>
        /// <response code="200">Successfully deleted.</response>
        /// <response code="400">One or more validation errors have occurred.</response>
        [HttpDelete("{ownerId}")]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
        public async Task<ActionResult> DeleteOwner([FromRoute] int ownerId, CancellationToken cancellationToken)
        {
            using (var transaction = new TransactionScope(TransactionScopeOption.Required,
                new TransactionOptions() { IsolationLevel = IsolationLevel.ReadCommitted }, TransactionScopeAsyncFlowOption.Enabled))
            {
                await _appService.DeleteOwner(ownerId);
                await _unitOfWork.SaveChangesAsync(cancellationToken);
                transaction.Complete();
            }
            return Ok();
        }

        /// <summary>
        /// </summary>
        /// <response code="200">Returns the specified List&lt;OwnerDTO&gt;.</response>
        /// <response code="400">One or more validation errors have occurred.</response>
        [HttpGet("*/lastname/{lastName}")]
        [ProducesResponseType(typeof(List<OwnerDTO>), StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
        public async Task<ActionResult<List<OwnerDTO>>> GetOwnersList(
            [FromRoute] string lastName,
            CancellationToken cancellationToken)
        {
            var result = default(List<OwnerDTO>);
            result = await _appService.GetOwnersList(lastName);
            return Ok(result);
        }


    }
}